src/main/java/org/elasticsearch/indices/recovery/RecoveriesCollection.java

```diff
@@ -128,42 +131,49 @@ public void markRecoveryAsDone(long id) {
         }
     }
 
-    /**
-     * Try to find an ongoing recovery for a given shard. returns null if not found.
-     */
-    @Nullable
-    public StatusRef findRecoveryByShard(IndexShard indexShard) {
-        for (RecoveryStatus recoveryStatus : onGoingRecoveries.values()) {
-            // check if the recovery has already finished and if not protect
-            // against it being closed on us while we check
-            if (recoveryStatus.tryIncRef()) {
-                try {
-                    if (recoveryStatus.indexShard() == indexShard) {
-                        recoveryStatus.incRef();
-                        return new StatusRef(recoveryStatus);
-                    }
-                } finally {
-                    recoveryStatus.decRef();
-                }
-            }
-        }
-        return null;
-    }
-
     /** the number of ongoing recoveries */
     public int size() {
         return onGoingRecoveries.size();
     }
 
     /** cancel all ongoing recoveries for the given shard. typically because the shards is closed */
-    public void cancelRecoveriesForShard(ShardId shardId, String reason) {
+    public boolean cancelRecoveriesForShard(ShardId shardId, String reason) {
+        return cancelRecoveriesForShard(shardId, reason, null);
+    }
+
+    /**
+     * cancel all ongoing recoveries for the given shard, if their status match a predicate
+     *
+     * @param reason       reason for cancellation
+     * @param shardId      shardId for which to cancel recoveries
+     * @param shouldCancel a predicate to check if a recovery should be cancelled or not. Null means cancel without an extra check.
+     *                     note that the recovery state can change after this check, but before it is being cancelled via other
+     *                     already issued outstanding references.
+     * @return true if a recovery was cancelled
+     */
+    public boolean cancelRecoveriesForShard(ShardId shardId, String reason, @Nullable Predicate<RecoveryStatus> shouldCancel) {
+        boolean cancelled = false;
         for (RecoveryStatus status : onGoingRecoveries.values()) {
             if (status.shardId().equals(shardId)) {
-                cancelRecovery(status.recoveryId(), reason);
+                boolean cancel = true;
+                if (shouldCancel != null) {
+                    if (status.tryIncRef()) {
+                        try {
+                            cancel = shouldCancel.apply(status);
+                        } finally {
+                            status.decRef();
+                        }
+                    }
+                }
+                if (cancel && cancelRecovery(status.recoveryId(), reason)) {
```

Comment:
doubting. In reality there is only one and it's unclear if the semantics of cancel anything you can and report errors in the end is worth the extra complexity. Let me know if you want this and I'll add the rethrow and suppress shenanigans